low-level/actuators/impl/generic/realstor_encl.py

```diff
@@ -608,7 +641,95 @@ def _get_encl_response(self, uri, request_type):
                 return response
             else:
                 logger.error(f"invalid data for {uri}")
-                return None
+                return {}
         except ValueError as err:
             logger.error(f"invalid data for {uri} {err}")
             return None
+
+    def _put_enclosure_action(self, ctrl_action, ctrl_type, resource,
+            enclosure_request):
+        severity = "informational"
+        message = "request performed successfully."
+        invalid_args = False
+        ctrl_cmd = ""
+        action_type = ""
+        if resource.startswith(("controller_a", "controller_b")):
+            ctrl_name = resource.split("controller_")[1].split(':')[0]
+        elif resource == "*":
+            ctrl_name = "both"
+        else:
+            ctrl_name = ""
+            invalid_args = True
+
+        if ctrl_action == "shutdown":
+            ctrl_cmd = f"{ctrl_action} {ctrl_name}"
+            action_type = ctrl_action
+        else:
+            ctrl_cmd = f"{ctrl_action} {ctrl_type} {ctrl_name}"
+            action_type = f"{ctrl_action} {ctrl_type}"
+
+        if (ctrl_action == "restart" and not resource.endswith(('sc', 'mc', "*"))) \
+            or (ctrl_action == "shutdown" and resource.endswith(('sc', 'mc'))) or \

```

Comment:
>or (ctrl_action == "shutdown" and resource.endswith(('sc', 'mc', **'*'**))) or \

Shouldn't * needs to be covered for shutdown, * is not a valid ctrl_name here

Reply:
I have written this condition for validate correct resource_id with correct controller action.
The meaning of the above condition is
> ctrl_action == "restart" and **not** resource.endswith(('sc', 'mc', "*")))  ===> Don't proceed

if **ctrl_action == "restart"** and resource is not ending with sc, mc or * then don't proceed further because **it should end with either sc, mc or "*"**

> ctrl_action == "shutdown" and resource.endswith(('sc', 'mc'))) ===> Don't proceed

**ctrl_action == "shutdown"** and resource is ending with sc or mc then don't proceed further because it is an **invalid resource type.** "*" will be allowed to shut down.